pallets/file-system/src/lib.rs

```diff
@@ -0,0 +1,290 @@
+        /// Because this pallet emits events, it depends on the runtime's definition of an event.
+        type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+        /// Type for identifying a file, generally a hash.
+        type Fingerprint: Parameter
+            + Member
+            + MaybeSerializeDeserialize
+            + MaybeDisplay
+            + SimpleBitOps
+            + Ord
+            + Default
+            + Copy
+            + CheckEqual
+            + AsRef<[u8]>
+            + AsMut<[u8]>
+            + MaxEncodedLen;
+
+        /// Unit representing the size of a file.
+        type StorageUnit: Parameter
+            + Member
+            + MaybeSerializeDeserialize
+            + Default
+            + MaybeDisplay
+            + AtLeast32Bit
+            + Copy
+            + MaxEncodedLen
+            + HasCompact;
+
+        /// Maximum number of BSPs that can store a file.
+        #[pallet::constant]
+        type MaxBspsPerStorageRequest: Get<u32>;
+
+        /// Maximum byte size of a file path.
+        #[pallet::constant]
+        type MaxFilePathSize: Get<u32>;
+
+        /// Maximum byte size of a libp2p multiaddress.
+        #[pallet::constant]
+        type MaxMultiAddressSize: Get<u32>;
+
+        /// Time-to-live for a storage request.
+        #[pallet::constant]
+        type StorageRequestTtl: Get<u32>;
+
+        /// Maximum number of expired storage requests to clean up in a single block.
+        #[pallet::constant]
+        type MaxExpiredStorageRequests: Get<u32>;
+    }
+
+    #[pallet::pallet]
+    pub struct Pallet<T>(_);
+
+    #[pallet::storage]
+    #[pallet::getter(fn storage_requests)]
+    pub type StorageRequests<T: Config> =
+        StorageMap<_, Blake2_128Concat, FileLocation<T>, StorageRequestMetadata<T>>;
+
+    #[pallet::storage]
+    #[pallet::getter(fn storage_requests_expirations)]
+    pub type StorageRequestExpirations<T: Config> = StorageMap<
+        _,
+        Blake2_128Concat,
+        BlockNumberFor<T>,
+        BoundedVec<FileLocation<T>, T::MaxExpiredStorageRequests>,
+    >;
+
+    #[pallet::event]
+    #[pallet::generate_deposit(pub(super) fn deposit_event)]
+    pub enum Event<T: Config> {
+        /// Notifies that a new file has been requested to be stored.
+        NewStorageRequest {
+            who: T::AccountId,
+            location: FileLocation<T>,
+            fingerprint: Fingerprint<T>,
+            size: StorageUnit<T>,
+            user_multiaddr: MultiAddress<T>,
+        },
+
+        /// Notifies that a BSP has been accepted to store a given file.
+        AcceptedBspVolunteer {
+            who: T::AccountId,
+            location: FileLocation<T>,
+            fingerprint: Fingerprint<T>,
+            bsp_multiaddress: MultiAddress<T>,
+        },
+
+        /// Notifies the expiration of a storage request.
+        StorageRequestExpired { location: FileLocation<T> },
```

Comment:
I'd also add an event for when a storage request is fully completed (and therefore deleted). Anyway, we'll have to add many events down the line.

Reply:
Yes definitely, I add the events when I use them, just to avoid having too much unused code